compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -864,6 +867,7 @@ public void visit(BLangMatchStatement matchStatement) {
                     containsLastPatternInStatement || (matchClause.matchGuard == null && this.hasLastPatternInClause);
         }
         this.statementReturns = allClausesReturns && containsLastPatternInStatement;
+        resetLastPatternInClause();
```

Comment:
Shouldn't we reset `this.hasLastPatternInClause` here to what was there before visiting the match clauses?

The following doesn't fail with this change.

```ballerina
function func(any a, any b) returns string {
    match a {
        _ => {
            match b {
                "bar" => {
                    return "s1";
                }
                _ => {
                    return "s2";
                }
            }
        }
        "foo" => { // should get an error
            
        }
    }
    return "";
}
```

Reply:
Fixed